projects/RabbitMQ.Client/ThrottlingRateLimiter.cs

```diff
@@ -0,0 +1,117 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright (c) 2007-2024 Broadcom. All Rights Reserved.
+//---------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.RateLimiting;
+using System.Threading.Tasks;
+
+namespace RabbitMQ.Client
+{
+    public class ThrottlingRateLimiter : RateLimiter
+    {
+        public const int DefaultThrottlingPercentage = 50;
+
+        private readonly ConcurrencyLimiter _concurrencyLimiter;
+        private readonly int _maxConcurrency;
+        private readonly int _throttlingThreshold;
+
+        public ThrottlingRateLimiter(int maxConcurrentCalls, int? throttlingPercentage = DefaultThrottlingPercentage)
+        {
+            _maxConcurrency = maxConcurrentCalls;
+            _throttlingThreshold = _maxConcurrency * throttlingPercentage.GetValueOrDefault(DefaultThrottlingPercentage) / 100;
+
+            ConcurrencyLimiterOptions limiterOptions = new()
+            {
+                QueueLimit = _maxConcurrency,
+                PermitLimit = _maxConcurrency
+            };
+
+            _concurrencyLimiter = new ConcurrencyLimiter(limiterOptions);
+        }
+
+        public override RateLimiterStatistics? GetStatistics() => _concurrencyLimiter.GetStatistics();
+
+        protected override RateLimitLease AttemptAcquireCore(int permitCount)
+        {
+            RateLimitLease lease = _concurrencyLimiter.AttemptAcquire(permitCount);
+
+            ThrottleIfNeeded();
+
+            return lease;
+        }
+
+        protected override async ValueTask<RateLimitLease> AcquireAsyncCore(int permitCount, CancellationToken cancellationToken)
+        {
+            RateLimitLease lease = await _concurrencyLimiter.AcquireAsync(permitCount, cancellationToken).ConfigureAwait(false);
+
+            await ThrottleIfNeededAsync(cancellationToken).ConfigureAwait(false);
+
+            return lease;
+        }
+
+        private void ThrottleIfNeeded()
+        {
+            long? availablePermits = _concurrencyLimiter.GetStatistics()?.CurrentAvailablePermits;
+            if (!(availablePermits < _throttlingThreshold))
+            {
+                return;
+            }
+
+            int delay = (int)((1.0 - availablePermits / (double)_maxConcurrency) * 1000);
+            Thread.Sleep(delay);
+        }
+
+        private Task ThrottleIfNeededAsync(CancellationToken cancellationToken = default)
+        {
+            long? availablePermits = _concurrencyLimiter.GetStatistics()?.CurrentAvailablePermits;
+            if (!(availablePermits < _throttlingThreshold))
+            {
+                return Task.CompletedTask;
+            }
+
+            int delay = (int)((1.0 - availablePermits / (double)_maxConcurrency) * 1000);
```

Comment:
Is it worth encapsulating this calculation into a method because we have it in two places?

Reply:
I can throw that in quickly, good idea.